### PR TITLE
Reuse Jenkins node and only install/build once

### PR DESCRIPTION
**Pull request by jamesdonoh**
Resolves https://github.com/bbc/simorgh-infrastructure/issues/1097

**Overall change:**
Prevents Jenkins from allocating a new `node` (and therefore workspace) for each Docker container, and to make this possible moves the install/build steps into a separate stage.

**Code changes**
- Tell Jenkins to use the top-level node for Docker containers to prevent additional workspaces from being created (`reuseNode true`)
- Move the installation of dependencies and build process into separate stages 
- Remove superfluous install/build steps to prevent interactions between parallel tests

**Context**
I believe the duplicate `@2`, `@3` etc. workspaces are created each time a Docker container is started in our pipeline. Jenkins reaches the `agent` directive in each stage and tries to allocate a new node. When it finds the default-named workspace in use by another node, it automatically duplicates it and appends `@NUM`. This allows each container to run with an isolated workspace directory. Example:

```
207M    simorgh_use-idt2-fallbacks-TJOHS2HB2WUXELTZCFCNMQBR2AIHGOAFWJSYEHEGZAIBC3TTV3ZA
445M    simorgh_use-idt2-fallbacks-TJOHS2HB2WUXELTZCFCNMQBR2AIHGOAFWJSYEHEGZAIBC3TTV3ZA@2
322M    simorgh_use-idt2-fallbacks-TJOHS2HB2WUXELTZCFCNMQBR2AIHGOAFWJSYEHEGZAIBC3TTV3ZA@3
```

However, the Jenkins [Workspace Cleanup plugin](https://www.jenkins.io/doc/pipeline/steps/ws-cleanup/) (`cleanWs`) that we invoke at the end lacks the ability to clean up these `@NUM` workspaces and only clears the original default-named workspace on the first node that is allocated (i.e. the top-level node in the pipeline).

I have arrived at this theory by running repeated builds of the same branch and observing how multiple workspace directories are created across different Jenkins nodes and never cleared up. However I can't find it officially documented anywhere. [This StackOverflow question](https://stackoverflow.com/questions/33282783/preserve-jenkins-workspace) claims that `agent` inside `stage` causes Jenkins to create "workspace@2" etc folders.

According to the [Jenkins docs](https://www.jenkins.io/doc/book/pipeline/syntax/#common-options), `reuseNode true`:
> [runs] the container on the node specified at the top-level of the Pipeline, in the same workspace, rather than on a new node entirely

This prevents the `@NUM` workspaces being created and runs everything in a single directory (on a single node).

**However:** In order for this to work, we need to ensure that the parallel tests we run (unit, e2e, etc.) do not interact with each other. In the current pipeline these independently run `npm ci` and `npm run build` etc. This PR consolidates those into separate *Install dependencies* and *Build* stages that run upfront before any of the tests start. Without this, one test suite running `npm run build` while another is running will cause the build to sporadically fail due to race conditions.

**Risks/disadvantages**
- This approach means that all the tests will still run in parallel but on a single Jenkins node, meaning that they may take longer to complete if the node is heavily loaded.
- There may still be other interactions and therefore race conditions between the parallel tests running in the same workspace that have not yet been identified

**Other links**
- [Example of `reuseNode`](https://github.com/jenkinsci/pipeline-model-definition-plugin/wiki/Controlling-your-build-environment#reusing-nodeworkspace-with-per-stage-docker-agents) in Jenkins docs

---

- [x] I have assigned myself to this PR and the corresponding issues
- [x] I have added labels to this PR for the relevant pod(s) affected by these changes
- [x] I have assigned this PR to the Simorgh project

**Testing:**

- [NA] Automated (jest and/or cypress) tests added (for new features) or updated (for existing features)
- [no] If necessary, I have run the local E2E non-smoke tests relevant to my changes (`CYPRESS_APP_ENV=local CYPRESS_SMOKE=false npm run test:e2e:interactive`)
- [no] This PR requires manual testing

### CODE REVIEW

File: package.json [hunk 1]
```diff
@@ -25,7 +25,8 @@
     "start": "NODE_ENV=production node build/server.js",
     "stop": "./scripts/killApp.sh",
     "storybook": "start-storybook -p 9001 -s .storybook/static -c .storybook",
-    "test": "npm run test:lint && npm run test:dependencies && npm run test:unit && npm run build -- --hide-modules --colors && run-p --race start amp:validate && run-p --race start test:integration:ci",
+    "test": "npm run build && npm run test:local",
+    "test:local": "npm run test:lint && npm run test:dependencies && npm run test:unit && run-p --race start amp:validate && run-p --race start test:integration:ci",
```
Comment (jamesdonoh) [inline package.json:29]: I've moved these around in order to preserve the current behaviour of `npm test` for those who rely upon it while allowing the tests to be run without triggering `build` again for CI use.

I don't know if `test:local` is a helpful name, but then to be honest I don't really understand the thinking behind the current split between `test` and `test:ci`.

File: Jenkinsfile [hunk 2]
```diff
@@ -228,11 +260,11 @@ pipeline {
           agent {
             docker {
               image "${nodeImage}"
+              reuseNode true
             }
           }
           steps {
             sh "rm -f storybook.zip"
-            sh 'make install'
```
Comment (jamesdonoh) [inline Jenkinsfile:235]: I can't see any reason why it should be necessary to install deps again at this point but happy to be corrected!

**Review by chris-hinds — APPROVED** [review]

**Review by amywalkerdev — CHANGES_REQUESTED** [review]
Running `npm test` aaaalll my tests are failing. Will see whats up

**Review by amywalkerdev — APPROVED** [review]
Nope seems to be an issue with my macbook, all run fine on my desktop 😄 
Great changes!

**Pull request merged**